### PR TITLE
Implement Soft Searching 

### PR DESCRIPTION
**Pull request by joycrwu**
## Overview

Added +/- 24 hour getting ride capability to the search feature

## Changes Made

- Added filtering by date time as well, including returning those that are within -24 hours of inputted departure time and + 24 hours of inputted departure time
- Deleted some comments from the previous PR
- Added the iexact for casing 

## Test Coverage

Manual testing with postman

Endpoint: http://127.0.0.1:8000/api/search/ -- GET

```
{
    "departure_datetime": "2020-03-27",
    "start_location_name": "home",
    "end_location_name": "test end"
}
```

```
[
    {
        "id": 1,
        "creator": {
            "id": 1,
            "netid": "test18",
            "first_name": "Tester",
            "last_name": "Test",
            "profile_pic_url": null,
            "grade": null,
            "pronouns": null
        },
        "max_travelers": 4,
        "min_travelers": 1,
        "departure_datetime": "2020-03-27T00:00:00Z",
        "description": "",
        "driver": null,
        "is_flexible": true,
        "riders": [],
        "estimated_cost": null,
        "path": {
            "id": 1,
            "start_location_place_id": "test13",
            "start_location_name": "home",
            "end_location_place_id": "test2",
            "end_location_name": "test end"
        },
        "type": "rideshare"
    },
    {
        "id": 2,
        "creator": {
            "id": 1,
            "netid": "test18",
            "first_name": "Tester",
            "last_name": "Test",
            "profile_pic_url": null,
            "grade": null,
            "pronouns": null
        },
        "max_travelers": 4,
        "min_travelers": 1,
        "departure_datetime": "2020-03-27T00:00:00Z",
        "description": "",
        "driver": null,
        "is_flexible": true,
        "riders": [],
        "estimated_cost": null,
        "path": {
            "id": 1,
            "start_location_place_id": "test13",
            "start_location_name": "home",
            "end_location_place_id": "test2",
            "end_location_name": "test end"
        },
        "type": "rideshare"
    },
    {
        "id": 3,
        "creator": {
            "id": 1,
            "netid": "test18",
            "first_name": "Tester",
            "last_name": "Test",
            "profile_pic_url": null,
            "grade": null,
            "pronouns": null
        },
        "max_travelers": 4,
        "min_travelers": 1,
        "departure_datetime": "2020-03-27T00:00:00Z",
        "description": "",
        "driver": null,
        "is_flexible": true,
        "riders": [],
        "estimated_cost": null,
        "path": {
            "id": 1,
            "start_location_place_id": "test13",
            "start_location_name": "home",
            "end_location_place_id": "test2",
            "end_location_name": "test end"
        },
        "type": "rideshare"
    },
    {
        "id": 4,
        "creator": {
            "id": 1,
            "netid": "test18",
            "first_name": "Tester",
            "last_name": "Test",
            "profile_pic_url": null,
            "grade": null,
            "pronouns": null
        },
        "max_travelers": 4,
        "min_travelers": 1,
        "departure_datetime": "2020-03-28T00:00:00Z",
        "description": "",
        "driver": null,
        "is_flexible": true,
        "riders": [],
        "estimated_cost": null,
        "path": {
            "id": 1,
            "start_location_place_id": "test13",
            "start_location_name": "home",
            "end_location_place_id": "test2",
            "end_location_name": "test end"
        },
        "type": "rideshare"
    }
]
```

## Related PRs or Issues (delete if not applicable)

Implement search feature -- can delete that PR tho

  

</details>

### CODE REVIEW

File: requirements.txt
Comment (chalo2000) [inline requirements.txt:2]: what's this?

File: src/ride/controllers/create_ride_controller.py
Comment (chalo2000) [inline src/ride/controllers/create_ride_controller.py]: see https://github.com/cuappdev/scoop-backend/pull/19#discussion_r983894284 for why we don't do this anymore

File: src/ride/controllers/search_ride_controller.py
Comment (chalo2000) [inline src/ride/controllers/search_ride_controller.py:25]: no need to pass an empty list, data default argument is None

File: src/ride/controllers/search_ride_controller.py
Comment (chalo2000) [inline src/ride/controllers/search_ride_controller.py]: instead of splitting, just do
`departure_datetime_object = datetime.datetime.fromisoformat(departure_datetime)`
https://docs.python.org/3/library/datetime.html#datetime.datetime.fromisoformat

File: src/ride/controllers/search_ride_controller.py
Comment (chalo2000) [inline src/ride/controllers/search_ride_controller.py]: remove

File: src/ride/views.py
Comment (chalo2000) [inline src/ride/views.py:63]: let's make a push to get rid of the try except and just load the json - I don't think we'll ever have a case where we have a request that is not a JSON.

**Review by chalo2000 — CHANGES_REQUESTED** [review]
some fixes + improvements before merging in

also, i noticed the datetime format is YYYY-MM-DD. if there are two departures in one day, how do I know which one is departing first? shouldn't we be considering time in this equation or is this related to the app design and I'm not aware of this? regardless of the answer, `fromisoformat` should take care of date strings with and without time so we should not manually be parsing them but please ensure that we do not need to include time in the departure and leave a comment on this PR when you've confirmed

File: requirements.txt
Comment (katesliang) [reply requirements.txt:2]: we needed to add this in order to setup the server on joyce's laptop

File: src/ride/controllers/create_ride_controller.py
Comment (katesliang) [inline src/ride/controllers/create_ride_controller.py]: remember to pull and merge from main to resolve all the merge conflicts!

File: src/ride/serializers.py
Comment (katesliang) [inline src/ride/serializers.py]: remove whitespace

File: src/ride/views.py
Comment (katesliang) [reply src/ride/views.py:63]: lol i took this from pear-django-backend; i don't think it hurts to have this reinforcement

File: src/ride/controllers/search_ride_controller.py
Comment (katesliang) [reply src/ride/controllers/search_ride_controller.py]: +1

File: src/ride/controllers/search_ride_controller.py
Comment (katesliang) [inline src/ride/controllers/search_ride_controller.py:38]: does this work if path is None, just curious

File: requirements.txt
Comment (joycrwu) [reply requirements.txt:2]: Added this to get my development environment working

**Review by katesliang — APPROVED** [review]
lgtm!! make sure to address gonzalo's comments, address merge conflicts, and close #19  :)

File: src/ride/controllers/create_ride_controller.py
Comment (joycrwu) [reply src/ride/controllers/create_ride_controller.py]: I think this change is from one of Kate's PR's

File: src/ride/controllers/search_ride_controller.py
Comment (joycrwu) [reply src/ride/controllers/search_ride_controller.py:25]: changed

File: src/ride/controllers/search_ride_controller.py
Comment (joycrwu) [reply src/ride/controllers/search_ride_controller.py]: also changed

File: src/ride/controllers/search_ride_controller.py
Comment (joycrwu) [reply src/ride/controllers/search_ride_controller.py]: changed x3

File: src/ride/views.py
Comment (joycrwu) [reply src/ride/views.py:63]: changed x4

**Comment by katesliang** [comment]
@chalo2000 the datetime format for the departure datetime is YYYY-MM-DDZ(plus the time) so we will know which rides are departing first

File: src/ride/controllers/create_ride_controller.py
Comment (chalo2000) [reply src/ride/controllers/create_ride_controller.py]: in that case make sure Kate changes this

File: src/ride/controllers/search_ride_controller.py
Comment (chalo2000) [reply src/ride/controllers/search_ride_controller.py:25]: still looks the same to me

File: src/ride/controllers/search_ride_controller.py
Comment (chalo2000) [reply src/ride/controllers/search_ride_controller.py]: ?

**Review by chalo2000 — APPROVED** [review]
I still don't see the changes in the resolved comments. I'm approving so you can get this in but make sure that you actually pushed the changes.

**Pull request merged**
